gensim/matutils.py

```diff
@@ -119,51 +119,42 @@ def corpus2csc(corpus, num_terms=None, dtype=np.float64, num_docs=None, num_nnz=
         Convert sparse format to Gensim corpus format.
 
     """
-    try:
-        # if the input corpus has the `num_nnz`, `num_docs` and `num_terms` attributes
-        # (as is the case with MmCorpus for example), we can use a more efficient code path
-        if num_terms is None:
-            num_terms = corpus.num_terms
-        if num_docs is None:
-            num_docs = corpus.num_docs
-        if num_nnz is None:
-            num_nnz = corpus.num_nnz
-    except AttributeError:
-        pass  # not a MmCorpus...
+    # if the input corpus has the `num_nnz`, `num_docs` and `num_terms` attributes
+    # (as is the case with MmCorpus for example), we can use a more efficient code path
+    if num_terms is None:
+        num_terms = getattr(corpus, 'num_terms', None)
+    if num_docs is None:
+        num_docs = getattr(corpus, 'num_docs', None)
+    if num_nnz is None:
+        num_nnz = getattr(corpus, 'num_nnz', None)
     if printprogress:
         logger.info("creating sparse matrix from corpus")
-    if num_terms is not None and num_docs is not None and num_nnz is not None:
-        # faster and much more memory-friendly version of creating the sparse csc
-        posnow, indptr = 0, [0]
-        indices = np.empty((num_nnz,), dtype=np.int32)  # HACK assume feature ids fit in 32bit integer
-        data = np.empty((num_nnz,), dtype=dtype)
-        for docno, doc in enumerate(corpus):
-            if printprogress and docno % printprogress == 0:
-                logger.info("PROGRESS: at document #%i/%i", docno, num_docs)
-            posnext = posnow + len(doc)
-            indices[posnow: posnext] = [feature_id for feature_id, _ in doc]
-            data[posnow: posnext] = [feature_weight for _, feature_weight in doc]
-            indptr.append(posnext)
-            posnow = posnext
-        assert posnow == num_nnz, "mismatch between supplied and computed number of non-zeros"
-        result = scipy.sparse.csc_matrix((data, indices, indptr), shape=(num_terms, num_docs), dtype=dtype)
-    else:
-        # slower version; determine the sparse matrix parameters during iteration
-        num_nnz, data, indices, indptr = 0, [], [], [0]
-        for docno, doc in enumerate(corpus):
-            if printprogress and docno % printprogress == 0:
-                logger.info("PROGRESS: at document #%i", docno)
-            indices.extend(feature_id for feature_id, _ in doc)
-            data.extend(feature_weight for _, feature_weight in doc)
-            num_nnz += len(doc)
-            indptr.append(num_nnz)
-        if num_terms is None:
-            num_terms = max(indices) + 1 if indices else 0
+    if num_nnz is None:
+        num_nnz = sum(len(doc) for doc in corpus)
```

Comment:
`corpus` may be (and often is) a generator. We cannot afford any extra passes like this. Make sure the refactoring keeps it single-pass.